File: hard/2147_number_of_ways_to_divide_a_long_corridor/solution.py
class Solution:
    def numberOfWays(self, corridor: str) -> int:
        section_start = -1
        prev_section_end = -1

        number_of_ways = 1

        for i, item in enumerate(corridor):
            if item != "S":
                continue

            if section_start == -1:
                section_start = i

                if prev_section_end != -1:
                    number_of_ways *= section_start - prev_section_end
            else:
                prev_section_end = i
                section_start = -1

        if section_start != -1:
            return 0
        elif prev_section_end == -1:
            return 0

        return number_of_ways % (10**9 + 7)

File: hard/2147_number_of_ways_to_divide_a_long_corridor/solution.cpp
using namespace std;

#include <string>

class Solution {
public:
  int numberOfWays(string corridor) {
    static const int MOD = 1'000'000'007;

    int sectionStart = -1;
    int prevSectionEnd = -1;

    long numberOfWays = 1;

    for (int i = 0; i < corridor.size(); ++i) {
      if (corridor[i] != 'S')
        continue;

      if (sectionStart == -1) {
        sectionStart = i;

        if (prevSectionEnd != -1)
          numberOfWays = (numberOfWays * (sectionStart - prevSectionEnd)) % MOD;
      } else {
        prevSectionEnd = i;
        sectionStart = -1;
      }
    }

    if (sectionStart != -1 || prevSectionEnd == -1)
      return 0;

    return numberOfWays;
  }
};
